crates/cxx-qt-lib/src/types/qset/mod.rs

```diff
@@ -0,0 +1,426 @@
+// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+use core::{marker::PhantomData, mem::MaybeUninit};
+use cxx::{type_id, ExternType};
+
+mod qset_bool;
+mod qset_f32;
+mod qset_f64;
+mod qset_i16;
+mod qset_i32;
+mod qset_i8;
+mod qset_qdate;
+mod qset_qdatetime;
+mod qset_qstring;
+mod qset_qtime;
+mod qset_qurl;
+mod qset_u16;
+mod qset_u32;
+mod qset_u8;
+
+// TODO: do we need packed?
+#[repr(C, packed)]
+pub struct QSet<T>
+where
+    T: QSetElement,
+{
+    _space: MaybeUninit<usize>,
+    _value: PhantomData<T>,
+}
+
+impl<T> Clone for QSet<T>
+where
+    T: QSetElement,
+{
+    /// Constructs a copy of the QSet.
+    fn clone(&self) -> Self {
+        T::clone(self)
+    }
+}
+
+impl<T> Default for QSet<T>
+where
+    T: QSetElement,
+{
+    /// Constructs an empty set.
+    fn default() -> Self {
+        T::default()
+    }
+}
+
+impl<T> Drop for QSet<T>
+where
+    T: QSetElement,
+{
+    /// Destroys the QSet.
+    fn drop(&mut self) {
+        T::drop(self);
+    }
+}
+
+impl<T> QSet<T>
+where
+    T: QSetElement,
+{
+    /// Removes all elements from the set.
+    pub fn clear(&mut self) {
+        T::clear(self);
+    }
+
+    /// Returns true if the set contains item value; otherwise returns false.
+    pub fn contains(&self, value: T) -> bool {
+        T::contains(self, value)
+    }
+
+    /// Returns true if the set contains no elements; otherwise returns false.
+    pub fn is_empty(&self) -> bool {
+        T::len(self) == 0
+    }
+
+    /// Returns the number of items in the set.
+    pub fn len(&self) -> isize {
+        T::len(self)
+    }
+}
+
+impl<T> QSet<T>
+where
+    T: QSetElement + ExternType<Kind = cxx::kind::Trivial>,
+{
+    /// Inserts item value into the set, if value isn't already in the set,
+    /// and returns an iterator pointing at the inserted item.
+    pub fn insert(&mut self, value: T) {
+        T::insert(self, value);
+    }
+
+    /// Removes any occurrence of item value from the set.
+    /// Returns true if an item was actually removed; otherwise returns false.
+    pub fn remove(&mut self, value: T) -> bool {
+        T::remove(self, value)
+    }
+}
+
+unsafe impl<T> ExternType for QSet<T>
+where
+    T: ExternType + QSetElement,
+{
+    type Id = T::TypeId;
+    type Kind = T::Kind;
+}
+
+pub trait QSetElement: Sized {
+    type TypeId;
+
+    fn clear(set: &mut QSet<Self>);
+    fn clone(set: &QSet<Self>) -> QSet<Self>;
+    fn contains(set: &QSet<Self>, value: Self) -> bool;
+    fn default() -> QSet<Self>;
+    fn drop(set: &mut QSet<Self>);
+    // TODO: we are passing as a const ref internally to C++,
+    // so do we need to have ManuallyDrop?
+    fn insert(set: &mut QSet<Self>, value: Self)
```

Comment:
do we need ManuallyDrop if we are passing as a const-ref internally, so we can safely let Rust drop the item?